wnfs/src/private/privateref.rs

```diff
@@ -1,149 +1,160 @@
 use std::fmt::Debug;
 
-use super::{AesKey, ContentKey, PrivateNodeHeader, RevisionKey};
-use crate::{FsError, HashOutput, Namefilter};
+use super::{PrivateNodeHeader, SnapshotKey, TemporalKey, KEY_BYTE_SIZE};
+use crate::{AesError, FsError, HashOutput, Namefilter};
+use aes_kw::KekAes256;
 use anyhow::Result;
-use rand_core::RngCore;
+use libipld::Cid;
 use serde::{de::Error as DeError, ser::Error as SerError, Deserialize, Serialize};
 
 //--------------------------------------------------------------------------------------------------
 // Type Definitions
 //--------------------------------------------------------------------------------------------------
 
-/// PrivateRef holds the information to fetch associated node from a private forest and decrypt it if it is present.
+/// PrivateRef holds the information to fetch a specific node from the private forest and decrypt it.
+///
+/// It also includes required key material to decrypt/encrypt any future revisions of the node it points to.
 #[derive(Clone, PartialEq, Eq)]
 pub struct PrivateRef {
-    /// Sha3-256 hash of saturated namefilter.
-    pub(crate) saturated_name_hash: HashOutput,
-    /// Sha3-256 hash of the ratchet key.
-    pub(crate) content_key: ContentKey,
-    /// Skip-ratchet-derived key.
-    pub(crate) revision_key: RevisionKey,
+    /// Sha3-256 hash of saturated namefilter. Used as the label for identifying revisions of PrivateNodes in the PrivateForest.
+    pub saturated_name_hash: HashOutput,
+    /// Skip-ratchet-derived key. Gives read access to the revision pointed to and any newer revisions.
+    pub temporal_key: TemporalKey,
+    /// CID that identifies the exact value in the multivalue.
+    pub content_cid: Cid,
 }
 
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub(crate) struct PrivateRefSerializable {
     #[serde(rename = "name")]
     pub(crate) saturated_name_hash: HashOutput,
-    #[serde(rename = "contentKey")]
-    pub(crate) content_key: ContentKey,
-    #[serde(rename = "revisionKey")]
-    pub(crate) revision_key: Vec<u8>,
+    #[serde(rename = "snapshotKey")]
+    pub(crate) snapshot_key: SnapshotKey,
+    #[serde(rename = "temporalKey")]
+    pub(crate) temporal_key: Vec<u8>,
+    #[serde(rename = "contentCid")]
+    pub(crate) content_cid: Cid,
+}
+
+/// A pointer to a specific revision in the private forest
+/// together with the TemporalKey to decrypt any of these
+/// revisions.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct RevisionRef {
+    /// Sha3-256 hash of saturated namefilter. Used as the label for private nodes in the private forest.
+    pub saturated_name_hash: HashOutput,
+    /// Skip-ratchet-derived key. Gives read access to the revision pointed to and any newer revisions.
+    pub temporal_key: TemporalKey,
 }
```

Comment:
This is fine but thinking of an alternative way of going about this. As in:

```rust
pub struct Ref<T> {
    pub saturated_name_hash: HashOutput,
    pub temporal_key: TemporalKey,
    pub t: T,
}

pub type PrivatRef = Ref<Cid>;
pub type RevisionRef = Ref<()>;

impl PrivateRef { /* ... */ }
impl RevisionRef { /* ... */ }
```

Reply:
Yeah true. Another way to slice the problem would be this:
```rs
pub struct PrivateRef {
  revision_ref: RevisionRef,
  content_cid: Cid,
}
```

There's probably more ways. I like to keep things simpler at the cost of some duplication. Would that be ok for you, too?

(There's also perhaps some changes coming with the `PrivateLink` abstraction eventually that *may* throw a wrench into abstractions we'd introduce now)